src/dbnode/storage/shard.go

```diff
@@ -2372,6 +2377,32 @@ func (s *dbShard) Snapshot(
 	}
 	s.RUnlock()
 
+	var (
+		needsSnapshot bool
+		// NB(bodu): This always defaults to false if the record does not exist.
+		emptySnapshotOnDisk = s.emptySnapshotOnDiskByTime[blockStart]
+	)
+	s.forEachShardEntry(func(entry *lookup.Entry) bool {
+		if !entry.Series.IsBufferEmptyAtBlockStart(blockStart) {
+			needsSnapshot = true
+			return false
+		}
+		return true
+	})
+	// Only terminate early when we would be over-writing an empty snapshot fileset on disk.
+	// TODO(bodu): We could bootstrap empty snapshot state in the bs path to avoid doing extra
+	// snapshotting work after a bootstrap since this cached state gets cleared.
+	if !needsSnapshot && emptySnapshotOnDisk {
+		return nil
+	}
+	// We're proceeding w/ the snaphot at this point but we know it will be empty or contain data that is recoverable
+	// from commit logs since we would have rotated commit logs before checking for flushable data.
+	if needsSnapshot {
+		s.emptySnapshotOnDiskByTime[blockStart] = false
+	} else {
+		s.emptySnapshotOnDiskByTime[blockStart] = true
+	}
```

Comment:
Seems like we should hold the series.Lock() (write lock()) while setting this data?

Reply:
I understand only called by a single caller but perhaps we write lock anyway given that this contract could change?

Reply:
Ah okay, mmm I guess it's fine in this case since this happens pretty fast and relatively infrequently so it won't increase shard lock contention by a significant amount.